src/models/mmt_encoder.py

```diff
@@ -0,0 +1,202 @@
+         units=hidden_size,
+         kernel_initializer=initializer,
+         name='patch_embedding_projection')
+
+     self._embedding_norm_layer = tf.keras.layers.LayerNormalization(
+         axis=-1, epsilon=1e-12, name='embeddings/layer_norm')
+
+     self._embedding_dropout_layer = tf.keras.layers.Dropout(
+         rate=hidden_dropout_prob, name='embeddings/dropout')
+
+     self._relative_transformer = etc_layers.RelativeTransformerLayers(
+         hidden_size=hidden_size,
+         num_hidden_layers=num_hidden_layers,
+         num_attention_heads=num_attention_heads,
+         intermediate_size=intermediate_size,
+         hidden_act=activation,
+         hidden_dropout_prob=hidden_dropout_prob,
+         attention_probs_dropout_prob=attention_probs_dropout_prob,
+         initializer_range=initializer_range,
+         relative_vocab_size=relative_vocab_size,
+         use_pre_activation_order=use_pre_activation_order,
+         use_one_hot_lookup=use_one_hot_lookup)
+     
+  def call(self,
+           word_ids: tf.Tensor,
+           segment_ids: Optional[tf.Tensor] = None,
+           att_mask: Optional[tf.Tensor] = None,
+           relative_att_ids: Optional[tf.Tensor] = None,
+           patch_embeddings: Optional[tf.Tensor] = None,
+           training: Optional[bool] = None):
+
+    """Calls the layer.
+      Args:
+        word_ids: <int32>[batch_size, seq_len] Tensor of word piece ids.
+        segment_ids: <int32>[batch_size, seq_len] Optional Tensor of segment
+          ids. By default we just fill all elements with segment id 1.
+        att_mask: <int32>[batch_size, seq_len,  seq_len].
+        relative_att_ids: <int32>[batch_size, seq_len, seq_len].
+        training: For Keras, optional boolean scalar tensor or Python boolean
+          indicating whether the call is meant for training or inference.
+      Returns:
+        <float32>[batch_size, seq_len, hidden_size].
+    """
+
+    if segment_ids is None:
+      segment_ids = tf.ones_like(word_ids, dtype=tf.int32)
+
+    word_embeddings = self._word_embedding_layer(word_ids)
+    segment_embeddings = self._segment_embedding_layer(segment_ids)
+
+    embeddings = word_embeddings + segment_embeddings
+
+    if self._position_embedding_layer is not None:
+      position_embeddings = self._position_embedding_layer(word_embeddings)
+      embeddings += position_embeddings
+
+    if patch_embeddings is not None:
+      seq_len = tf_utils.get_shape_list(word_embeddings)[1]
+      patch_seq_len = tf_utils.get_shape_list(patch_embeddings)[1]
+
+      patch_embeddings = self._patch_embedding_projection(patch_embeddings)
+
+      # Make patch_embeddings and word_embeddings have the same shape.
+      # 2 is for CLS and [PATCH]
+      prefix_pad_len = 2
+      suffix_pad_len = seq_len - 2 - patch_seq_len 
+      patch_embeddings = tf.pad(
+          patch_embeddings,
+          paddings=[[0, 0], [prefix_pad_len, suffix_pad_len], [0, 0]])
+      embeddings += patch_embeddings
```

Comment:
Nit: consider adding unitest for this. Index and dim/shape are easily getting messed up during coding.

Reply:
+1

Reply:
will do!